File: csharp/examples/calculator/calculator.cs
using System.Globalization;
using Uimd;

internal static class CalculatorConstants
{
    public const string DisplayZero = "0";
    public const string DisplayError = "Error";
    public const double PercentFactor = 100.0;
}

internal sealed class ExpressionParser
{
    private readonly string input;
    private int position;

    public ExpressionParser(string input)
    {
        this.input = input;
    }

    public double Parse()
    {
        double value = Expression();
        SkipSpaces();
        if (position != input.Length)
        {
            throw new InvalidOperationException("unexpected trailing calculator input");
        }
        return value;
    }

    private double Expression()
    {
        double value = Term();
        while (true)
        {
            SkipSpaces();
            if (Match('+'))
            {
                value += Term();
            }
            else if (Match('-'))
            {
                value -= Term();
            }
            else
            {
                return value;
            }
        }
    }

    private double Term()
    {
        double value = Factor();
        while (true)
        {
            SkipSpaces();
            if (Match('*'))
            {
                value *= Factor();
            }
            else if (Match('/'))
            {
                value /= Factor();
            }
            else
            {
                return value;
            }
        }
    }

    private double Factor()
    {
        SkipSpaces();
        if (Match('+'))
        {
            return Factor();
        }
        if (Match('-'))
        {
            return -Factor();
        }
        return Number();
    }

    private double Number()
    {
        SkipSpaces();
        int start = position;
        while (position < input.Length && (char.IsDigit(input[position]) || input[position] == '.'))
        {
            ++position;
        }
        if (start == position)
        {
            throw new InvalidOperationException("expected calculator number");
        }
        return double.Parse(input[start..position], CultureInfo.InvariantCulture);
    }

    private bool Match(char value)
    {
        if (position < input.Length && input[position] == value)
        {
            ++position;
            return true;
        }
        return false;
    }

    private void SkipSpaces()
    {
        while (position < input.Length && char.IsWhiteSpace(input[position]))
        {
            ++position;
        }
    }
}

public sealed class Calculator : CalculatorUI
{
    private string expression = CalculatorConstants.DisplayZero;
    private bool afterResult;

    public Calculator()
    {
        Refresh();
    }

    public override GeneratedWindowRuntimeOptions RuntimeOptions()
    {
        GeneratedWindowRuntimeOptions options = base.RuntimeOptions();
        options.OnKey = Key;
        return options;
    }

    protected override void onN0Click()
    {
        Button("n0");
    }

    protected override void onN1Click()
    {
        Button("n1");
    }

    protected override void onN2Click()
    {
        Button("n2");
    }

    protected override void onN3Click()
    {
        Button("n3");
    }

    protected override void onN4Click()
    {
        Button("n4");
    }

    protected override void onN5Click()
    {
        Button("n5");
    }

    protected override void onN6Click()
    {
        Button("n6");
    }

    protected override void onN7Click()
    {
        Button("n7");
    }

    protected override void onN8Click()
    {
        Button("n8");
    }

    protected override void onN9Click()
    {
        Button("n9");
    }

    protected override void onDtClick()
    {
        Button("dt");
    }

    protected override void onAdClick()
    {
        Button("ad");
    }

    protected override void onMnClick()
    {
        Button("mn");
    }

    protected override void onMlClick()
    {
        Button("ml");
    }

    protected override void onDvClick()
    {
        Button("dv");
    }

    protected override void onSbClick()
    {
        Button("sb");
    }

    protected override void onAcClick()
    {
        Button("ac");
    }

    protected override void onDlClick()
    {
        Button("dl");
    }

    protected override void onPrClick()
    {
        Button("pr");
    }

    private void Button(string name)
    {
        switch (name)
        {
            case "n0": Digit('0'); break;
            case "n1": Digit('1'); break;
            case "n2": Digit('2'); break;
            case "n3": Digit('3'); break;
            case "n4": Digit('4'); break;
            case "n5": Digit('5'); break;
            case "n6": Digit('6'); break;
            case "n7": Digit('7'); break;
            case "n8": Digit('8'); break;
            case "n9": Digit('9'); break;
            case "dt": Digit('.'); break;
            case "ad": Operation('+'); break;
            case "mn": Operation('-'); break;
            case "ml": Operation('*'); break;
            case "dv": Operation('/'); break;
            case "sb": EqualsPressed(); break;
            case "ac": Reset(); break;
            case "dl": Delete(); break;
            case "pr": Percent(); break;
        }
    }

    private bool Key(string key)
    {
        if (key.Length == 1 && char.IsDigit(key[0]))
        {
            Digit(key[0]);
            return true;
        }
        if (key == ".")
        {
            Digit('.');
            return true;
        }
        if (key is "+" or "-" or "*" or "/")
        {
            Operation(key[0]);
            return true;
        }
        if (key == "=")
        {
            EqualsPressed();
            return true;
        }
        if (key == "%")
        {
            Percent();
            return true;
        }
        if (key == "Backspace")
        {
            Delete();
            return true;
        }
        return false;
    }

    private void Digit(char digit)
    {
        if (afterResult || expression == CalculatorConstants.DisplayError)
        {
            expression = "";
            afterResult = false;
        }

        if (digit == '.')
        {
            AppendDecimalPoint();
        }
        else if (expression == CalculatorConstants.DisplayZero)
        {
            expression = digit.ToString();
        }
        else if (CurrentNumberText() is "0" or "-0")
        {
            ReplaceCurrentNumber(CurrentNumberText().Replace(CalculatorConstants.DisplayZero, digit.ToString(), StringComparison.Ordinal));
        }
        else
        {
            expression += digit;
        }
        Refresh();
    }

    private void Operation(char operation)
    {
        if (expression == CalculatorConstants.DisplayError)
        {
            expression = CalculatorConstants.DisplayZero;
        }

        if (expression.Length == 0)
        {
            if (operation == '-')
            {
                expression = operation.ToString();
                afterResult = false;
                Refresh();
            }
            return;
        }

        if (IsOperator(expression[^1]))
        {
            if (operation == '-' && expression[^1] != '-')
            {
                expression += operation;
            }
            else
            {
                expression = expression.TrimEnd('+', '-', '*', '/') + operation;
            }
        }
        else
        {
            expression += operation;
        }
        afterResult = false;
        Refresh();
    }

    private void EqualsPressed()
    {
        Compute();
    }

    private void Delete()
    {
        if (afterResult || expression == CalculatorConstants.DisplayError)
        {
            Reset();
            return;
        }
        expression = expression.Length > 0 ? expression[..^1] : "";
        if (expression.Length == 0)
        {
            expression = CalculatorConstants.DisplayZero;
        }
        Refresh();
    }

    private void Percent()
    {
        try
        {
            string current = CurrentNumberText();
            if (current.Length == 0)
            {
                return;
            }
            ReplaceCurrentNumber(Format(double.Parse(current, CultureInfo.InvariantCulture) / CalculatorConstants.PercentFactor));
            afterResult = false;
            Refresh();
        }
        catch (Exception)
        {
            ShowError();
        }
    }

    private void Compute()
    {
        try
        {
            double result = new ExpressionParser(expression).Parse();
            if (!double.IsFinite(result))
            {
                throw new InvalidOperationException("non-finite calculator result");
            }
            expression = Format(result);
            afterResult = true;
            Refresh();
        }
        catch (Exception)
        {
            ShowError();
        }
    }

    private void Reset()
    {
        expression = CalculatorConstants.DisplayZero;
        afterResult = false;
        Refresh();
    }

    private void Refresh()
    {
        display.SetText(string.IsNullOrEmpty(expression) ? CalculatorConstants.DisplayZero : expression);
    }

    private void ShowError()
    {
        expression = CalculatorConstants.DisplayError;
        afterResult = true;
        Refresh();
    }

    private void AppendDecimalPoint()
    {
        string current = CurrentNumberText();
        if (current.Contains('.', StringComparison.Ordinal))
        {
            return;
        }
        if (expression.Length == 0 || IsOperator(expression[^1]))
        {
            expression += CalculatorConstants.DisplayZero + ".";
        }
        else
        {
            expression += ".";
        }
    }

    private (int Start, int End) CurrentNumberSpan()
    {
        int end = expression.Length;
        if (end == 0 || IsOperator(expression[end - 1]))
        {
            return (end, end);
        }

        int start = end;
        while (start > 0 && (char.IsDigit(expression[start - 1]) || expression[start - 1] == '.'))
        {
            --start;
        }
        if (start > 0 && expression[start - 1] == '-')
        {
            if (start == 1 || IsOperator(expression[start - 2]))
            {
                --start;
            }
        }
        return (start, end);
    }

    private string CurrentNumberText()
    {
        (int start, int end) = CurrentNumberSpan();
        return expression[start..end];
    }

    private void ReplaceCurrentNumber(string value)
    {
        (int start, int end) = CurrentNumberSpan();
        expression = expression[..start] + value + expression[end..];
    }

    private static bool IsOperator(char value)
    {
        return value is '+' or '-' or '*' or '/';
    }

    private static string Format(double value)
    {
        double rounded = Math.Round(value);
        if (Math.Abs(value - rounded) < double.Epsilon * Math.Max(1.0, Math.Abs(value)))
        {
            return rounded.ToString("0", CultureInfo.InvariantCulture);
        }
        return value.ToString("G10", CultureInfo.InvariantCulture);
    }
}

public static class Program
{
    public static int Main(string[] args)
    {
        return GeneratedWindowRuntime.RunGeneratedAppMain(() =>
        {
            Calculator app = new();
            return GeneratedWindowRuntime.RunGeneratedWindow(app, app.RuntimeOptions(), args);
        });
    }
}

File: java/examples/calculator/Calculator.java
import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;

import uimd.GeneratedWindowRuntime;
import uimd.GeneratedWindowRuntimeOptions;

final class CalculatorConstants
{
    static final String DISPLAY_ZERO = "0";
    static final String DISPLAY_ERROR = "Error";
    static final double PERCENT_FACTOR = 100.0;

    private CalculatorConstants()
    {
    }
}

final class ExpressionParser
{
    private final String input;
    private int position;

    ExpressionParser(String input)
    {
        this.input = input;
    }

    double parse()
    {
        double value = expression();
        skipSpaces();
        if (position != input.length())
        {
            throw new IllegalStateException("unexpected trailing calculator input");
        }
        return value;
    }

    private double expression()
    {
        double value = term();
        while (true)
        {
            skipSpaces();
            if (match('+'))
            {
                value += term();
            }
            else if (match('-'))
            {
                value -= term();
            }
            else
            {
                return value;
            }
        }
    }

    private double term()
    {
        double value = factor();
        while (true)
        {
            skipSpaces();
            if (match('*'))
            {
                value *= factor();
            }
            else if (match('/'))
            {
                value /= factor();
            }
            else
            {
                return value;
            }
        }
    }

    private double factor()
    {
        skipSpaces();
        if (match('+'))
        {
            return factor();
        }
        if (match('-'))
        {
            return -factor();
        }
        return number();
    }

    private double number()
    {
        skipSpaces();
        int start = position;
        while (position < input.length()
            && (Character.isDigit(input.charAt(position)) || input.charAt(position) == '.'))
        {
            ++position;
        }
        if (start == position)
        {
            throw new IllegalStateException("expected calculator number");
        }
        return Double.parseDouble(input.substring(start, position));
    }

    private boolean match(char value)
    {
        if (position < input.length() && input.charAt(position) == value)
        {
            ++position;
            return true;
        }
        return false;
    }

    private void skipSpaces()
    {
        while (position < input.length() && Character.isWhitespace(input.charAt(position)))
        {
            ++position;
        }
    }
}

public final class Calculator extends CalculatorUI
{
    private static final MathContext DISPLAY_PRECISION = new MathContext(10, RoundingMode.HALF_EVEN);

    private String expression = CalculatorConstants.DISPLAY_ZERO;
    private boolean afterResult;

    public Calculator()
    {
        refresh();
    }

    @Override
    public GeneratedWindowRuntimeOptions runtimeOptions()
    {
        GeneratedWindowRuntimeOptions options = super.runtimeOptions();
        options.setOnKey(this::key);
        return options;
    }

    @Override
    protected void onN0Click()
    {
        button("n0");
    }

    @Override
    protected void onN1Click()
    {
        button("n1");
    }

    @Override
    protected void onN2Click()
    {
        button("n2");
    }

    @Override
    protected void onN3Click()
    {
        button("n3");
    }

    @Override
    protected void onN4Click()
    {
        button("n4");
    }

    @Override
    protected void onN5Click()
    {
        button("n5");
    }

    @Override
    protected void onN6Click()
    {
        button("n6");
    }

    @Override
    protected void onN7Click()
    {
        button("n7");
    }

    @Override
    protected void onN8Click()
    {
        button("n8");
    }

    @Override
    protected void onN9Click()
    {
        button("n9");
    }

    @Override
    protected void onDtClick()
    {
        button("dt");
    }

    @Override
    protected void onAdClick()
    {
        button("ad");
    }

    @Override
    protected void onMnClick()
    {
        button("mn");
    }

    @Override
    protected void onMlClick()
    {
        button("ml");
    }

    @Override
    protected void onDvClick()
    {
        button("dv");
    }

    @Override
    protected void onSbClick()
    {
        button("sb");
    }

    @Override
    protected void onAcClick()
    {
        button("ac");
    }

    @Override
    protected void onDlClick()
    {
        button("dl");
    }

    @Override
    protected void onPrClick()
    {
        button("pr");
    }

    private void button(String name)
    {
        switch (name)
        {
            case "n0": digit('0'); break;
            case "n1": digit('1'); break;
            case "n2": digit('2'); break;
            case "n3": digit('3'); break;
            case "n4": digit('4'); break;
            case "n5": digit('5'); break;
            case "n6": digit('6'); break;
            case "n7": digit('7'); break;
            case "n8": digit('8'); break;
            case "n9": digit('9'); break;
            case "dt": digit('.'); break;
            case "ad": operation('+'); break;
            case "mn": operation('-'); break;
            case "ml": operation('*'); break;
            case "dv": operation('/'); break;
            case "sb": equalsPressed(); break;
            case "ac": reset(); break;
            case "dl": delete(); break;
            case "pr": percent(); break;
            default: break;
        }
    }

    private boolean key(String key)
    {
        if (key.length() == 1 && Character.isDigit(key.charAt(0)))
        {
            digit(key.charAt(0));
            return true;
        }
        if (".".equals(key))
        {
            digit('.');
            return true;
        }
        if (key.length() == 1 && isOperator(key.charAt(0)))
        {
            operation(key.charAt(0));
            return true;
        }
        if ("=".equals(key))
        {
            equalsPressed();
            return true;
        }
        if ("%".equals(key))
        {
            percent();
            return true;
        }
        if ("Backspace".equals(key))
        {
            delete();
            return true;
        }
        return false;
    }

    private void digit(char digit)
    {
        if (afterResult || CalculatorConstants.DISPLAY_ERROR.equals(expression))
        {
            expression = "";
            afterResult = false;
        }

        if (digit == '.')
        {
            appendDecimalPoint();
        }
        else if (CalculatorConstants.DISPLAY_ZERO.equals(expression))
        {
            expression = Character.toString(digit);
        }
        else if ("0".equals(currentNumberText()) || "-0".equals(currentNumberText()))
        {
            replaceCurrentNumber(currentNumberText().replace(
                CalculatorConstants.DISPLAY_ZERO,
                Character.toString(digit)));
        }
        else
        {
            expression += digit;
        }
        refresh();
    }

    private void operation(char operation)
    {
        if (CalculatorConstants.DISPLAY_ERROR.equals(expression))
        {
            expression = CalculatorConstants.DISPLAY_ZERO;
        }

        if (expression.isEmpty())
        {
            if (operation == '-')
            {
                expression = Character.toString(operation);
                afterResult = false;
                refresh();
            }
            return;
        }

        if (isOperator(expression.charAt(expression.length() - 1)))
        {
            if (operation == '-' && expression.charAt(expression.length() - 1) != '-')
            {
                expression += operation;
            }
            else
            {
                expression = trimTrailingOperators(expression) + operation;
            }
        }
        else
        {
            expression += operation;
        }
        afterResult = false;
        refresh();
    }

    private void equalsPressed()
    {
        compute();
    }

    private void delete()
    {
        if (afterResult || CalculatorConstants.DISPLAY_ERROR.equals(expression))
        {
            reset();
            return;
        }
        if (!expression.isEmpty())
        {
            expression = expression.substring(0, expression.length() - 1);
        }
        if (expression.isEmpty())
        {
            expression = CalculatorConstants.DISPLAY_ZERO;
        }
        refresh();
    }

    private void percent()
    {
        try
        {
            String current = currentNumberText();
            if (current.isEmpty())
            {
                return;
            }
            replaceCurrentNumber(format(
                Double.parseDouble(current) / CalculatorConstants.PERCENT_FACTOR));
            afterResult = false;
            refresh();
        }
        catch (RuntimeException exception)
        {
            showError();
        }
    }

    private void compute()
    {
        try
        {
            double result = new ExpressionParser(expression).parse();
            if (!Double.isFinite(result))
            {
                throw new IllegalStateException("non-finite calculator result");
            }
            expression = format(result);
            afterResult = true;
            refresh();
        }
        catch (RuntimeException exception)
        {
            showError();
        }
    }

    private void reset()
    {
        expression = CalculatorConstants.DISPLAY_ZERO;
        afterResult = false;
        refresh();
    }

    private void refresh()
    {
        display.setText(expression.isEmpty() ? CalculatorConstants.DISPLAY_ZERO : expression);
    }

    private void showError()
    {
        expression = CalculatorConstants.DISPLAY_ERROR;
        afterResult = true;
        refresh();
    }

    private void appendDecimalPoint()
    {
        String current = currentNumberText();
        if (current.indexOf('.') >= 0)
        {
            return;
        }
        if (expression.isEmpty() || isOperator(expression.charAt(expression.length() - 1)))
        {
            expression += CalculatorConstants.DISPLAY_ZERO + ".";
        }
        else
        {
            expression += ".";
        }
    }

    private int[] currentNumberSpan()
    {
        int end = expression.length();
        if (end == 0 || isOperator(expression.charAt(end - 1)))
        {
            return new int[]{end, end};
        }

        int start = end;
        while (start > 0
            && (Character.isDigit(expression.charAt(start - 1)) || expression.charAt(start - 1) == '.'))
        {
            --start;
        }
        if (start > 0 && expression.charAt(start - 1) == '-'
            && (start == 1 || isOperator(expression.charAt(start - 2))))
        {
            --start;
        }
        return new int[]{start, end};
    }

    private String currentNumberText()
    {
        int[] span = currentNumberSpan();
        return expression.substring(span[0], span[1]);
    }

    private void replaceCurrentNumber(String value)
    {
        int[] span = currentNumberSpan();
        expression = expression.substring(0, span[0]) + value + expression.substring(span[1]);
    }

    private static String trimTrailingOperators(String value)
    {
        int end = value.length();
        while (end > 0 && isOperator(value.charAt(end - 1)))
        {
            --end;
        }
        return value.substring(0, end);
    }

    private static boolean isOperator(char value)
    {
        return value == '+' || value == '-' || value == '*' || value == '/';
    }

    private static String format(double value)
    {
        return BigDecimal.valueOf(value)
            .round(DISPLAY_PRECISION)
            .stripTrailingZeros()
            .toPlainString();
    }

    private static void require(boolean condition, String message)
    {
        if (!condition)
        {
            throw new IllegalStateException(message);
        }
    }

    private static void smoke()
    {
        CalculatorUI window = new CalculatorUI();
        require("Calculator".equals(window.title()), "calculator title mismatch");
        require(window.generatedLayout().size() == 20, "calculator layout mismatch");
        require(window.display != null, "calculator display is missing");
        require(window.n0 != null && "0".equals(window.n0.title()), "calculator digit button mismatch");
        require(window.sb != null && "=".equals(window.sb.title()), "calculator equals button mismatch");
    }

    private static void logicTest()
    {
        Calculator application = new Calculator();
        require(application.key("1"), "calculator key 1 was not handled");
        require(application.key("2"), "calculator key 2 was not handled");
        require(application.key("+"), "calculator key + was not handled");
        require(application.key("3"), "calculator key 3 was not handled");
        require(application.key("="), "calculator key = was not handled");
        require("15".equals(application.display.text()), "calculator addition failed");
        application.button("ml");
        application.button("n2");
        application.button("sb");
        require("30".equals(application.display.text()), "calculator multiplication failed");
        application.button("pr");
        require("0.3".equals(application.display.text()), "calculator percent failed");
        application.button("ac");
        require("0".equals(application.display.text()), "calculator reset failed");
    }

    public static void main(String[] arguments)
    {
        if (arguments.length > 0 && "--smoke".equals(arguments[0]))
        {
            smoke();
            return;
        }
        if (arguments.length > 0 && "--logic-test".equals(arguments[0]))
        {
            logicTest();
            return;
        }
        int exitCode = GeneratedWindowRuntime.runGeneratedAppMain(() ->
        {
            Calculator application = new Calculator();
            return GeneratedWindowRuntime.runGeneratedWindow(
                application,
                application.runtimeOptions(),
                arguments);
        });
        System.exit(exitCode);
    }
}
